clang/test/CIR/CodeGen/AArch64/neon.c

```diff
@@ -9617,12 +9617,16 @@ poly16x8_t test_vmull_p8(poly8x8_t a, poly8x8_t b) {
 //   return vqaddh_s16(a, b);
 // }
 
-// NYI-LABEL: @test_vqadds_s32(
-// NYI:   [[VQADDS_S32_I:%.*]] = call i32 @llvm.aarch64.neon.sqadd.i32(i32 %a, i32 %b)
-// NYI:   ret i32 [[VQADDS_S32_I]]
-// int32_t test_vqadds_s32(int32_t a, int32_t b) {
-//   return vqadds_s32(a, b);
-// }
+int32_t test_vqadds_s32(int32_t a, int32_t b) {
+  return vqadds_s32(a, b);
+
+  // CIR: vqadds_s32
+  // CIR: cir.binop(add, {{%.*}}, {{%.*}}) nsw nuw sat : !s32i
```

Comment:
Seems like nsw nuw could also co-exist with sat, please update the helpers I mentioned in one of the previous comments.

Reply:
This is just CIR attribute though so we decide whether they can co-exist, but when it lowers to LLVM IR, since saturated BinOp will just become llvm intrinsic call, so nsw  and nuw not applicable there.

Reply:
I don't understand, if it's not applicable why it shows up in the test?

> This is just CIR attribute though so we decide whether they can co-exist

This is my whole point, you should tell me whether they should co-exist or not and enforce whatever semantics are right for the job

Reply:
> I don't understand, if it's not applicable why it shows up in the test?
> 
> > This is just CIR attribute though so we decide whether they can co-exist
> 
> This is my whole point, you should tell me whether they should co-exist or not and enforce whatever semantics are right for the job

I see. Let's not make them co-exist for now. On one hand, arithmetically , saturated value is not wrap-around value 
But my concern is that if a future opt pass would treat overflow value of an Op with nsw and nuw as Poison value as it should, then it certainly makes no sense when saturated flag is on, as the result would not be a poison value, it's intended to be saturated value.

Reply:
Sounds good. Btw, what does OG codegen generate here?

Reply:
> Sounds good. Btw, what does OG codegen generate here?

[OG uses `llvm.aarch64.neon.sqadd.i32`, intrinsic.](https://godbolt.org/z/Mrf5h4con).

Reply:
That's right, so you'd never see the nsw/nuw thing, gotcha